desc/magnetic_fields/_core.py

```diff
@@ -540,6 +540,75 @@ def save_mgrid(
         file.close()
 
 
+class MagneticFieldFromUser(_MagneticField, Optimizable):
+    """Wrap an arbitrary function for calculating magnetic field in lab coordinates.
+
+    Parameters
+    ----------
+    fun : callable
+        Function to compute magnetic field at arbitrary points. Should have a signature
+        of the form ``fun(coords, params) -> B`` where
+
+          - ``coords`` is a (n,3) array of positions in R, phi, Z coordinates where
+            the field is to be evaluated.
+          - ``params`` is an array of optional parameters, eg for optimizing the field.
+          - ``B`` is the returned value of the magnetic field as a (n,3) array in R,
+            phi, Z coordinates.
+
+    params : ndarray, optional
+        Default values for parameters. Defaults to an empty array.
+
+    """
+
+    def __init__(self, fun, params=None):
+        errorif(not callable(fun), ValueError, "fun must be callable")
+        self._params = setdefault(params, jnp.array([]))
+        self._fun = fun
```

Comment:
Would it be worth also checking that that `fun(coords, params)` returns an array of the expected shape for some dummy coordinates? So that the user doesn't get a mysterious error when they try to do `MagneticFieldFromUser.compute_magnetic_field(coords)`